### PR TITLE
Fix EZP-27462: Avoid duplication of path part of endpoint URL

### PR DESCRIPTION
**Pull request by natanael89**
[EZP-27462](https://jira.ez.no/browse/EZP-27462)
# 
Excerpt from the JIRA issue:
> When accessing the backoffice using URL with "app.php" (for example http://ezplatform.dev/app.php/ez) it is not possible to log in. The user is immediately redirected to login page and there is an error showing in the console: "_ezcombo?app-base/app-base-min.js&querystring-stringify-simple/querystring-stringify-simple-min.js&…:1687 POST http://ezplatform.dev/app.php/app.php/api/ezp/v2/user/sessions/r04atonm3r2i8n7l630r8o66ik/refresh 404 (Not Found)". Also, when you click the "Login" button, the login page displays an error "An unexpected error occurred" and there is an error in the console: "_ezcombo?app-base/app-base-min.js&querystring-stringify-simple/querystring-stringify-simple-min.js&…:1687 POST http://ezplatform.dev/app.php/app.php/api/ezp/v2/user/sessions 404 (Not Found)".
The URLs of these requests have the "app.php" two times, which means it is duplicated. Removing the duplication should resolve the bug.

The problem comes from how CAPI works.
Object ConnectionManager accepts among other things the "endPointUrl". The value of the "endPointUrl" equals "/app.php" in the example from the JIRA issue. There is also a test case in which the"endPointUrl" equals "http://ez.git.local". I figured that CAPI should accept the URLs with the origin part, as well as without it.
The "endPointUrl" is then used as a prefix for generated requests. If "endPointUrl" doesn't have anything in its path part, then everything works as expected. If "endPointURL" does have something in its path part, then requests generated completely from the information in CAPI work as expected, but some requests are generated from the previous requests' responses. Such requests are prefixed with "endPointUrl" (with path part), but the paths in responses from the server already have the full path added (in the example from the JIRA issue, they are already prefixed with "/app.php"). This means that after ConnectionManager prefixes it with the full "endPointUrl" the "/app.php" part is duplicated and therefore the request returns 404.
If the CAPI is to be used as it is currently, I think the ConnectionManager should accept only the origin part of the "endPointUrl". The path part should be used as a prefix for the "rootPath", which is then used to generate requests. If the requests are generated from the previous requests' responses, then the "rootPath" is not used.
For extracting the origin and path parts from the URL I used the following library: https://github.com/medialize/URI.js. Alternatively, I could extract from it only the parts I need, but I think that adding it as a dependency is a better option.

Another way to fix this problem would be to call the CAPI with different parameters (which means the fix would go to a place that is calling the CAPI), but I think this library should handle it on its own.

### CODE REVIEW

**Comment by andrerom** [comment]
@natanael89 If you and @adamwojs are confident in this change, and @dew326 don't see any issues with this for 1.x or 2.x. Then we can try to get it into 1.13.

File: test/CAPI.tests.js [hunk 1]
```diff
@@ -39,140 +43,185 @@ define(function (require) {
             this.setCredentials = function (credentials) { };
         };
 
-        beforeEach(function () {
-            testOptions = new TestOptions();
-            mockAuthenticationAgent = new MockAuthenticationAgent();
+        testUrls.forEach(function(testUrl)
+        {
```
Comment (dew326) [inline test/CAPI.tests.js]: I believe this is PHP CS. This `{` should in one line with `function(testUrl)`

File: test/CAPI.tests.js [hunk 2]
```diff
@@ -39,140 +43,185 @@ define(function (require) {
...
+            describe("for every url", function(){
```
Comment (dew326) [inline test/CAPI.tests.js]: Missing space between `()` and `{`

File: test/CAPI.tests.js [hunk 3]
```diff
@@ -39,140 +43,185 @@ define(function (require) {
...
+                        expect(capi).toBeDefined();
+                        expect(capi instanceof CAPI).toBeTruthy();
+                    });
+
+                    it("should update authenticationAgent with a reference to CAPI", function () {
+                        expect(mockAuthenticationAgent._CAPI).toBe(capi);
+                    });
+                });
+                describe("Service API", function () {
+                    describe('getContentervice', function(){
```
Comment (dew326) [inline test/CAPI.tests.js]: `() {` this is valid for a couple of other places in the code.

**Comment by dew326** [comment]
About 2.x I don't see any problem. We pass empty string as the `endPointUrl` so from my understanding the urijs will return empty strings for `pathname` and `origin`.

Please be aware the master is now 2.0 and the PlatformUIAssets are fetching only 1.x. We have to decide if we will release 1.6 (and merge it up to 2.0) or 2.1 (and allow PlatformUIAssets to fetch 2.x).

**Comment by natanael89** [comment]
Rebased and fixed @dew326's suggestions. I also moved the ticket to QA.

**Comment by andrerom** [comment]
@natanael89 can you rebase for 1.5 branch?

**Comment by natanael89** [comment]
@andrerom done

**Pull request merged**